### ISSUE TITLE
Find Distance at Which Value Changes by X%?

### ISSUE THREAD

**Issue by patrickcgray** [pysal/esda#125]
Hello, I've been really impressed with the pysal library and especially the documentation I've found in the ongoing Geographic Data Science Book: https://geographicdata.science/book/.

But I have a question that hasn't been addressed in the resources I've found. I have a set of polygons and each polygon has a continuous value from 0-1. I'm looking to find the average distance at which this value typically changes by a certain amount, let's say 10%. Is there a clean way to do this in pysal? If pysal doesn't have this functionality is there a name for this type of problem so that I research it further?

Basically what I'm hoping to do is say, even if I don't have data in a given location, how far away can a polygon with data be for me to still count it as the value for a different location.

**Comment by ljwolf** [comment]
Reflecting on this problem, I think there may be some data structures we can [adapt to provide this functionality easily](https://github.com/pysal/esda/files/12093742/1-s2.0-003132039090057R-main.pdf). I wonder, either, if the [conga line could be adapted here](https://github.com/carsonfarmer/fastpair/blob/master/fastpair/base.py). The basic API would be something like:

```python
def spatial_cumulant(locations, values, threshold, return_values=True):
    """
    For each location, find the distance `d` at which `reduction(values)` at `locations` closer than `d` exceeds `threshold`. 
    """
    tree = build_incremental_knn_structure(locations)
    complete = numpy.zeros_like(locations).astype(bool)
    k_vector = numpy.zeros_like(locations).astype(int)
    cumulant = numpy.zeros_like(locations).astype(values.dtype)
    while not complete.all():
        d_near, ix_near = tree.next_nearest(locations, k=k_vector)
        cumulant += values[ix_near] 
        complete |= cumulant > threshold
        k_vector += 1 - complete.astype(int)
    if return_values: 
        return d, cumulant
    return d 
```
where the `next_nearest()` function returns the distance and index to *just* the `k_vector`th neighbour. If a `cached_query()` function were available that returned *all k* efficiently (i.e. without re-doing the work for all k-1)*, we could then implement arbitrary reductions (like, `numpy.median()`), rather than just allowing for sum.

**Comment by ljwolf** [comment]
Such a data structure is in [`cgal`](https://doc.cgal.org/latest/Spatial_searching/classCGAL_1_1Orthogonal__incremental__neighbor__search.html), which includes [swig bindings](https://github.com/CGAL/cgal-swig-bindings/tree/main) that cover [incremental spatial search](https://doc.cgal.org/latest/Manual/packages.html#PkgSpatialSearching), but I don't see these wrapped by [`scikit-geometry`](https://github.com/scikit-geometry/scikit-geometry) yet...
